### DISCUSSION TITLE
Parsing Data

### DISCUSSION THREAD

**Discussion by ryanbooker** [pointfreeco/swift-parsing#20] [Q&A] [answered]
I'm trying to parse some well structured `Data` into a `String`. i.e. a `Parser<Data, String>`.

Basics: The `Data` contains `hex` bytes, some of which need to be "checked", then the remainder up to an "end" character, read and converted to ascii characters. 

I have already parsed it manually, and now I'd like to compare it to an equivalent `Parser`.

Any guidance on where to begin? Pointers to examples, etc?

Thanks in advance.

**Comment by stephencelis** [comment, marked as answer]
@ryanbooker We actually have a binary data parser in our benchmark suite: https://github.com/pointfreeco/swift-parsing/blob/dd2e6f2e8e7ca0211e81bf790044a314602fd7dd/Sources/swift-parsing-benchmark/BinaryData.swift

**Reply by ryanbooker** [reply]
Brilliant thanks! I missed that. :)

**Reply by GeekAndDad** [reply]
This is a useful example.  Thanks for including it.

Seems like it might be out of date now?  Or else I've just been away from programming Swift and using SwiftParsing too long (definitely true).

I made something like this but the 32 bit version:

```
struct Word16Parser: Parser {
  typealias Input = Data
  typealias Output = UInt16

  func parse(_ input: inout Data) -> UInt16? {
    guard input.count >= 2
    else { return nil }
    let output = UInt16(input[input.startIndex]) + UInt16(input[input.startIndex + 1]) << 8
    input.removeFirst(2)
    return output
  }
}
```

and I get a compiler error on the struct definition saying that `Word16Parser` does't conform to Parser (and it appears to be the `UInt16?` being optional that's the issue (?)).

Converting to `func parse(_ input: inout Data) throws -> UInt16` and changing the `return(nil)` to throw an error allowed me to move forward.

Cheers
